openprescribing/api/views_measures.py

```diff
@@ -79,141 +79,132 @@ def measure_global(request, format=None):
-        # For other measures, the columns used to calculate the numerator is
-        # not available here (it's in BQ) so we order by total_items, which is
-        # the best we can do.
-        #
-        # But because the columns in BQ don't match the columns in PG (items vs
-        # total_items), and because we alias a column in the query below
-        # (actual_cost vs cost) we need to translate the name of the column we
-        # use for ordering the results.
-        match = re.match(
-            "SUM\((items|quantity|actual_cost)\) AS numerator", m.numerator_columns
-        )
-
-        if match:
-            order_col = {
-                "items": "total_items",
-                "actual_cost": "cost",
-                "quantity": "quantity",
-            }[match.groups()[0]]
-        else:
-            order_col = "total_items"
-
-        # The redundancy in the following column names is so we can
-        # support various flavours of `WHERE` clause from the measure
-        # definitions that may use a subset of any of these column
-        # names
-        focus_on_org = org_id and org_field
-        params = {
-            "numerator_bnf_codes": m.numerator_bnf_codes,
-            "three_months_ago": three_months_ago,
-        }
-        if focus_on_org:
-            org_condition = "{org_field} = %(org_id)s AND ".format(org_field=org_field)
-            org_group = "{org_field}, ".format(org_field=org_field)
-            params["org_id"] = org_id
-        else:
-            org_condition = ""
-            org_group = ""
-        query = """
-            SELECT
-              presentation_code AS bnf_code,
-              pn.name AS presentation_name,
-              SUM(total_items) AS total_items,
-              SUM(actual_cost) AS cost,
-              SUM(quantity) AS quantity
-            FROM
-              frontend_prescription p
-            INNER JOIN
-              frontend_presentation pn
-            ON p.presentation_code = pn.bnf_code
-            {extra_join}
-            WHERE
-              {org_condition}
-              processing_date >= %(three_months_ago)s
-              AND
-              pn.bnf_code = ANY(%(numerator_bnf_codes)s)
-            GROUP BY
-              {org_group}
-              presentation_code, pn.name
-            ORDER BY {order_col} DESC
-            LIMIT 50
-        """.format(
-            org_condition=org_condition,
-            org_group=org_group,
-            org_field=org_field,
-            three_months_ago=three_months_ago,
-            extra_join=extra_join,
-            order_col=order_col,
+    measure_id = request.query_params.get("measure", None)
+    measure = Measure.objects.get(pk=measure_id)
+    org_type, org_id = _get_org_type_and_id_from_request(request)
+    group_by_org = get_row_grouper(org_type)
+
+    # Closure over `group_by_org` and `org_id` which takes a prescribing matrix
+    # and returns the total value for that organisation over the last N months
```

Comment:
N = 3.

Reply:
Pedant point - not actually a closure, just a nested function referencing a nonlocal variable?

Reply:
Ooh.  What's the difference?

Reply:
If we returned a reference to this function outside the current scope it would become a closure. Pedantry should never be apologised for.